architecture.adoc

```diff
@@ -0,0 +1,134 @@
+= Architecture
+
+= Introduction
+
+Mokey is an application that collects data about running applications on an Aurora OpenShift cluster and exposes that
+information via a ReST API in a cached and fast-to-access manner.
+
+The collection process is scheduled and parallelized, and information about the applications is updated in real time as
+the process progresses.
+
+This document is a description of the overall architecture of the module, divided into two main sections;
+<<core-concepts, the Core Concepts>> and <<api-resources, the API resources>>, describing the internal aspects of the
+module, and the externally exposed resources, respectively.
+
+
+[[core-concepts]]
+= Core Concepts (Internal)
+
+This section contains a description of the core concepts of the module, with implementation and domain model details.
+It does not contain any details of the resources that are exposed from the API, only the classes and objects that are
+used to support the internal data collection workflow; crawling.
+
+[[reference-data]]
+== Reference Data
+
+Classes and objects that represent static data and rules.
+
+Package: no.skatteetaten.aurora.mokey.model
+
+[[non-functional-requirement]]NonFunctionalRequirement (NFR):: A description of a non functional requirement for applications running on Aurora
+OpenShift.
+
+NfrViolationCause:: Maybe declare causes for NFR violations?
+
+NfrViolationFix:: Maybe declare fixes for NFR violations?
+
+
+[[application-data]]
+== Application Data
+
+All pieces of relevant data for an application based on the needs of the clients of the API. The data is collected from
+the Aurora OpenShift cluster, or calculated from it. Intended as an internal representation to be used as a base for
+external resources.
+
+Package: no.skatteetaten.aurora.mokey.model
+
+[[collected-data]]
+=== Collected Data
+
+Classes that represent data collected from the cluster for an application.
+
+ApplicationId:: name and Environment
+Environment:: name and an affiliation.
+ApplicationData:: Container object for all data collected for a given application. Gives access to calculated entities
+via properties.
```

Comment:
The collected data must be represented as s Try. So either a success or a Failure. Where a failure is modled as an error code and a message. 

This makes it possible to check why something failed.

An instance running the Application must have 
Try|Health, Try|Info, Try|Management

Resource endpoints from the application should be modled as entities not as JsonNode to ensure they are correct. 

The Info resource must be lax in its parsing of git/build info.